File: src/lib/tensorkv/serve.ts
import { TensorKVAppliance, TensorKVContext } from "./appliance";
import { PagedEngine } from "./engine";
import { Arrival, ServingScheduler } from "./scheduler";
import { ShareGPTWorkload, SplitMix64, TOKENS_PER_BLOCK } from "./types";

function percentile(xs: number[], p: number) {
  if (!xs.length) return 0;
  const ys = [...xs].sort((a, b) => a - b);
  const i = Math.min(ys.length - 1, Math.max(0, Math.round((p / 100) * (ys.length - 1))));
  return ys[i];
}

function summary(xs: number[]) {
  if (!xs.length) return { n: 0, mean: 0, p50: 0, p90: 0, p99: 0, max: 0 };
  return {
    n: xs.length,
    mean: xs.reduce((a, b) => a + b, 0) / xs.length,
    p50: percentile(xs, 50),
    p90: percentile(xs, 90),
    p99: percentile(xs, 99),
    max: Math.max(...xs),
  };
}

export function sharegptArrivals(opts: {
  nSessions?: number;
  nPrefixes?: number;
  prefixBlocks?: number;
  uniqueBlocks?: number;
  decodeTokens?: number;
  seed?: number;
  tpb?: number;
} = {}) {
  const nSessions = opts.nSessions ?? 16;
  const nPrefixes = opts.nPrefixes ?? 4;
  const prefixBlocks = opts.prefixBlocks ?? 10;
  const uniqueBlocks = opts.uniqueBlocks ?? 2;
  const decodeTokens = opts.decodeTokens ?? 3;
  const seed = opts.seed ?? 5;
  const tpb = opts.tpb ?? TOKENS_PER_BLOCK;
  const wl = new ShareGPTWorkload({
    nPrefixes,
    nSessions,
    prefixBlocks,
    uniqueBlocks,
    seed,
  });
  const rng = new SplitMix64(BigInt(seed + 17));
  const arrivals: Arrival[] = [];
  let t = 0;
  for (const s of wl.sessions) {
    const prefix = Array.from({ length: s.prefixBlocks * tpb }, (_, i) => s.prefixId * 256 + i);
    const suffix = Array.from({ length: s.uniqueBlocks * tpb }, (_, i) => 1_000_000 + s.sessionId * 256 + i);
    t += rng.randint(0, 2);
    arrivals.push({
      reqId: s.sessionId + 1,
      tokens: prefix.concat(suffix),
      prefixTokens: prefix,
      maxNew: Math.max(2, decodeTokens),
      arriveTick: t,
    });
  }
  return { wl, arrivals };
}

export function runServing(opts: {
  nSessions?: number;
  nPrefixes?: number;
  prefixBlocks?: number;
  uniqueBlocks?: number;
  maxBatch?: number;
  decodeTokens?: number;
  nPages?: number;
  nBuckets?: number;
  seed?: number;
} = {}) {
  const nSessions = opts.nSessions ?? 16;
  const nPrefixes = opts.nPrefixes ?? 4;
  const prefixBlocks = opts.prefixBlocks ?? 10;
  const uniqueBlocks = opts.uniqueBlocks ?? 2;
  const maxBatch = opts.maxBatch ?? 6;
  const decodeTokens = opts.decodeTokens ?? 3;
  const seed = opts.seed ?? 5;
  const nPages = Math.max(opts.nPages ?? 192, 32);
  const nBuckets = opts.nBuckets ?? 64;
  const { wl, arrivals } = sharegptArrivals({
    nSessions,
    nPrefixes,
    prefixBlocks,
    uniqueBlocks,
    decodeTokens,
    seed,
  });
  const tkv = new TensorKVAppliance({
    nBuckets,
    nPages,
    storePayloads: false,
    seed,
  });
  const eng = new PagedEngine(new TensorKVContext(tkv));
  const sch = new ServingScheduler(eng, maxBatch);
  const stats = sch.runArrivals(arrivals);
  const st = tkv.stats();
  return {
    nSessions,
    nPrefixes,
    prefixBlocks,
    uniqueBlocks,
    workingSetBlocks: wl.workingSetBlocks,
    nPages,
    maxBatch,
    submitted: stats.submitted,
    finished: stats.finished,
    prefixHits: stats.prefixHits,
    prefixHitRate: stats.submitted ? stats.prefixHits / stats.submitted : 0,
    decodeSteps: stats.decodeSteps,
    reclaims: stats.reclaims,
    ticks: stats.ticks,
    peakBatch: stats.peakBatch,
    vectorOverflow: stats.vectorOverflow,
    ttft: summary(stats.ttftMs),
    tbt: summary(stats.tbtMs),
    overflowBytes: eng.tkv.overflowBytes,
    descriptorsPosted: eng.tkv.descriptorsPosted,
    device: {
      hashLoad: st.hashLoad,
      pagesUsed: st.pagesUsed,
      getOps: st.gets,
      putOps: st.puts,
      probeOps: st.probes,
      evictOps: st.evicts,
      gatheredBytes: st.gatheredBytes,
    },
    complete: stats.finished === nSessions,
    ttftSeries: stats.ttftMs.map((ms, i) => ({ req: i + 1, ttft: Number(ms.toFixed(3)) })),
    getLatencyNs: summary(tkv.getLatencies),
  };
}

File: python/tensorkv/serve.py
"""ShareGPT-style continuous-batch serving on the TensorKV software stack.

Arrivals are Zipf-popular system prompts plus unique suffixes. The scheduler
PROBE/prefill/decode path issues real TKV_PUT / TKV_GET / TKV_PROBE / TKV_EVICT
on TensorKVAppliance. GET P99 in the report is the appliance histogram, not a
constant.
"""

from __future__ import annotations

from dataclasses import asdict

from .appliance import ApplianceConfig, TensorKVAppliance
from .constants import TOKENS_PER_BLOCK
from .engine import PagedEngine
from .hashutil import SplitMix64
from .libtkv import TensorKVContext
from .metrics import summary
from .scheduler import Arrival, ServingScheduler
from .workload import ShareGPTWorkload


def sharegpt_arrivals(
    *,
    n_sessions: int = 32,
    n_prefixes: int = 6,
    prefix_blocks: int = 10,
    unique_blocks: int = 2,
    decode_tokens: int = 4,
    seed: int = 5,
    tpb: int = TOKENS_PER_BLOCK,
) -> tuple[ShareGPTWorkload, list[Arrival]]:
    wl = ShareGPTWorkload(
        n_prefixes=n_prefixes,
        n_sessions=n_sessions,
        prefix_blocks=prefix_blocks,
        unique_blocks=unique_blocks,
        seed=seed,
    )
    rng = SplitMix64(seed + 17)
    arrivals: list[Arrival] = []
    t = 0
    for s in wl.sessions:
        prefix = list(range(s.prefix_id * 256, s.prefix_id * 256 + s.prefix_blocks * tpb))
        suffix = list(
            range(
                1_000_000 + s.session_id * 256,
                1_000_000 + s.session_id * 256 + s.unique_blocks * tpb,
            )
        )
        t += rng.randint(0, 2)
        arrivals.append(
            Arrival(
                req_id=s.session_id + 1,
                tokens=prefix + suffix,
                prefix_tokens=prefix,
                max_new=max(2, decode_tokens),
                arrive_tick=t,
            )
        )
    return wl, arrivals


def run_serving(
    n_sessions: int = 32,
    n_prefixes: int = 6,
    prefix_blocks: int = 10,
    unique_blocks: int = 2,
    max_batch: int = 8,
    decode_tokens: int = 4,
    n_pages: int = 256,
    n_buckets: int = 128,
    seed: int = 5,
) -> dict:
    wl, arrivals = sharegpt_arrivals(
        n_sessions=n_sessions,
        n_prefixes=n_prefixes,
        prefix_blocks=prefix_blocks,
        unique_blocks=unique_blocks,
        decode_tokens=decode_tokens,
        seed=seed,
    )
    pages = max(n_pages, 32)
    tkv = TensorKVAppliance(
        ApplianceConfig(n_buckets=n_buckets, n_pages=pages, store_payloads=False, seed=seed)
    )
    eng = PagedEngine(TensorKVContext(tkv))
    sch = ServingScheduler(eng, max_batch=max_batch)
    stats = sch.run_arrivals(arrivals)
    st = tkv.stats()
    get_lat = [float(x) for x in tkv.get_latencies]
    prefix_hit_rate = (stats.prefix_hits / stats.submitted) if stats.submitted else 0.0
    return {
        "n_sessions": n_sessions,
        "n_prefixes": n_prefixes,
        "prefix_blocks": prefix_blocks,
        "unique_blocks": unique_blocks,
        "working_set_blocks": wl.working_set_blocks,
        "n_pages": pages,
        "max_batch": max_batch,
        "submitted": stats.submitted,
        "finished": stats.finished,
        "prefix_hits": stats.prefix_hits,
        "prefix_hit_rate": prefix_hit_rate,
        "decode_steps": stats.decode_steps,
        "reclaims": stats.reclaims,
        "ticks": stats.ticks,
        "peak_batch": stats.peak_batch,
        "vector_overflow": stats.vector_overflow,
        "ttft": summary(stats.ttft_ms),
        "tbt": summary(stats.tbt_ms),
        "get_latency_ns": summary(get_lat),
        "overflow_bytes": eng.tkv.overflow_bytes,
        "descriptors_posted": eng.tkv.descriptors_posted,
        "device": {
            "hash_load": st["hash_load"],
            "pages_used": st["pages_used"],
            "get_ops": st["gets"],
            "put_ops": st["puts"],
            "probe_ops": st["probes"],
            "evict_ops": st["evicts"],
            "gathered_bytes": st["gathered_bytes"],
            "slow_inserts": st["slow_inserts"],
        },
        "engine": asdict(eng.stats),
        "complete": stats.finished == n_sessions,
    }


def serving_report(**kwargs) -> dict:
    return run_serving(**kwargs)
